crates/nu-command/src/core_commands/do_.rs

```diff
@@ -95,87 +107,89 @@ impl Command for Do {
-                            exit_code_stream.into_iter().collect()
-                        }
-                    };
-                    if let Some(Value::Int { val: code, .. }) = exit_code.last() {
-                        // if exit_code is not 0, it indicates error occured, return back Err.
-                        if *code != 0 {
-                            return Err(ShellError::ExternalCommand(
-                                "External command runs to failed".to_string(),
-                                stderr_msg,
-                                span,
-                            ));
-                        }
+        match result {
+            Ok(PipelineData::ExternalStream {
+                stdout,
+                stderr,
+                exit_code,
+                span,
+                metadata,
+            }) if capture_errors => {
+                // collect all output first.
+                let mut stderr_ctrlc = None;
+                let stderr_msg = match stderr {
+                    None => "".to_string(),
+                    Some(stderr_stream) => {
+                        stderr_ctrlc = stderr_stream.ctrlc.clone();
+                        stderr_stream.into_string().map(|s| s.item)?
                     }
-                    // construct pipeline data to our caller
-                    Ok(PipelineData::ExternalStream {
-                        stdout: Some(RawStream::new(
-                            Box::new(vec![Ok(stdout_msg.into_bytes())].into_iter()),
-                            stdout_ctrlc,
-                            span,
-                        )),
-                        stderr: Some(RawStream::new(
-                            Box::new(vec![Ok(stderr_msg.into_bytes())].into_iter()),
-                            stderr_ctrlc,
+                };
+
+                let mut stdout_ctrlc = None;
+                let stdout_msg = match stdout {
+                    None => "".to_string(),
+                    Some(stdout_stream) => {
+                        stdout_ctrlc = stdout_stream.ctrlc.clone();
+                        stdout_stream.into_string().map(|s| s.item)?
+                    }
+                };
+
+                let mut exit_code_ctrlc = None;
+                let exit_code: Vec<Value> = match exit_code {
+                    None => vec![],
+                    Some(exit_code_stream) => {
+                        exit_code_ctrlc = exit_code_stream.ctrlc.clone();
+                        exit_code_stream.into_iter().collect()
+                    }
+                };
+                if let Some(Value::Int { val: code, .. }) = exit_code.last() {
+                    // if exit_code is not 0, it indicates error occured, return back Err.
+                    if *code != 0 {
+                        return Err(ShellError::ExternalCommand(
+                            "External command failed".to_string(),
+                            stderr_msg,
```

Comment:
Does `capture-errors` imply that we're not ignoring program errors? I'm finding the code a bit hard to understand, but I don't think that's anything to do with your changes. However if you wouldn't mind factoring out some helper functions with helpful names I think that would help a lot.

Reply:
Yes, that's my understanding. The example is `do -c { nu -c 'exit 1' } | myscarycommand` where myscarycommand isn't supposed to run if things failed.